libraries/entities-renderer/src/EntityTreeRenderer.cpp

```diff
@@ -158,79 +158,86 @@ render::ItemID EntityTreeRenderer::renderableIdForEntityId(const EntityItemID& i
-            _entitiesScriptEngine->callEntityScriptMethod(entityID, "hoverOverEntity", event);
-        });
-        connect(entityScriptingInterface.data(), &EntityScriptingInterface::hoverLeaveEntity, _entitiesScriptEngine.data(), [&](const EntityItemID& entityID, const PointerEvent& event) {
-            _entitiesScriptEngine->callEntityScriptMethod(entityID, "hoverLeaveEntity", event);
-        });
-
-        connect(_entitiesScriptEngine.data(), &ScriptEngine::entityScriptPreloadFinished, [&](const EntityItemID& entityID) {
-            EntityItemPointer entity = getTree()->findEntityByID(entityID);
-            if (entity) {
-                entity->setScriptHasFinishedPreload(true);
-            }
-        });
+    entityScriptingInterface->setPersistentEntitiesScriptEngine(entitiesScriptEngineProvider);
 
-        _mouseAndPreloadSignalHandlersConnected = true;
-    }
+    setupEntityScriptEngineSignals(_persistentEntitiesScriptEngine);
+}
+
+void EntityTreeRenderer::resetNonPersistentEntitiesScriptEngine() {
+    _nonPersistentEntitiesScriptEngine = scriptEngineFactory(ScriptEngine::ENTITY_CLIENT_SCRIPT, NO_SCRIPT,
+                                                QString("about:Entities %1").arg(++_entitiesScriptEngineCount));
+    DependencyManager::get<ScriptEngines>()->runScriptInitializers(_nonPersistentEntitiesScriptEngine);
+    _nonPersistentEntitiesScriptEngine->runInThread();
+    auto entitiesScriptEngineProvider = qSharedPointerCast<EntitiesScriptEngineProvider>(_nonPersistentEntitiesScriptEngine);
+    DependencyManager::get<EntityScriptingInterface>()->setNonPersistentEntitiesScriptEngine(entitiesScriptEngineProvider);
+
+    setupEntityScriptEngineSignals(_nonPersistentEntitiesScriptEngine);
 }
 
 void EntityTreeRenderer::stopDomainAndNonOwnedEntities() {
     leaveDomainAndNonOwnedEntities();
     // unload and stop the engine
-    if (_entitiesScriptEngine) {
-        QList<EntityItemID> entitiesWithEntityScripts = _entitiesScriptEngine->getListOfEntityScriptIDs();
+    if (_nonPersistentEntitiesScriptEngine) {
+        QList<EntityItemID> entitiesWithEntityScripts = _nonPersistentEntitiesScriptEngine->getListOfEntityScriptIDs();
 
-        foreach (const EntityItemID& entityID,  entitiesWithEntityScripts) {
+        foreach (const EntityItemID& entityID, entitiesWithEntityScripts) {
             EntityItemPointer entityItem = getTree()->findEntityByEntityItemID(entityID);
-
             if (entityItem && !entityItem->getScript().isEmpty()) {
                 if (!(entityItem->isLocalEntity() || entityItem->isMyAvatarEntity())) {
-                    if (_currentEntitiesInside.contains(entityID)) {
-                        _entitiesScriptEngine->callEntityScriptMethod(entityID, "leaveEntity");
-                    }
-                    _entitiesScriptEngine->unloadEntityScript(entityID, true);
+                    _nonPersistentEntitiesScriptEngine->unloadEntityScript(entityID, true);
```

Comment:
Is it valid to remove the `_currentEntitiesInside` test here? Is `_currententitiesIside` still used OK elsewhere?

Reply:
leaveDomainAndNonOwnedEntities already calls leaveEntity on the things we are inside so we shouldn’t need to do it here